File: cpp/446_number_of_arithmetic_slices/number_of_arithmetic_slices.cpp
/*
 * @Date: 2021-08-11 14:38:50
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-08-11 14:43:41
 */

#include <cassert>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
 public:
  int numberOfArithmeticSlices(vector<int> &nums) {
    int ans = 0;
    int n = nums.size();
    vector<unordered_map<long long, int>> f(n);
    for (int i = 0; i < n; ++i) {
      for (int j = 0; j < i; ++j) {
        long long d = 1LL * nums[i] - nums[j];
        auto it = f[j].find(d);
        int cnt = it == f[j].end() ? 0 : it->second;
        ans += cnt;
        f[i][d] += cnt + 1;
      }
    }
    return ans;
  }
};

int main() {
  {
    vector<int> nums{2, 4, 6, 8, 10};
    assert(Solution{}.numberOfArithmeticSlices(nums) == 7);
  }
  {
    vector<int> nums{7, 7, 7, 7, 7};
    assert(Solution{}.numberOfArithmeticSlices(nums) == 16);
  }
}

File: rust/413_number_of_arithmetic_slices/number_of_arithmetic_slices.rs
/*
 * @Date: 2021-08-10 19:50:01
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-08-10 20:09:45
 */

struct Solution;

impl Solution {
    pub fn number_of_arithmetic_slices(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        if n == 1 {
            return 0;
        }

        let mut d = nums[0] - nums[1];
        let (mut t, mut ans) = (0, 0);

        for i in 2..n {
            if nums[i - 1] - nums[i] == d {
                t += 1;
            } else {
                d = nums[i - 1] - nums[i];
                t = 0;
            }
            ans += t;
        }
        ans
    }
}

fn main() {
    {
        let nums = vec![1, 2, 3, 4];
        assert_eq!(Solution::number_of_arithmetic_slices(nums), 3);
    }
    {
        let nums = vec![1];
        assert_eq!(Solution::number_of_arithmetic_slices(nums), 0);
    }
}
